File: client/lib/engine/team_assigner.dart
import 'types.dart';

Map<String, String> assignTeams(Map<String, List<GameCard>> hands) {
  final teams = <String, String>{};
  final redThreeHolders = <String>[];

  for (final entry in hands.entries) {
    final hasRedThree = entry.value.any((c) =>
        (c.suit == Suit.H || c.suit == Suit.D) && c.rank == Rank.THREE);
    if (hasRedThree) redThreeHolders.add(entry.key);
  }

  for (final playerId in hands.keys) {
    teams[playerId] = redThreeHolders.contains(playerId) ? 'red' : 'black';
  }
  return teams;
}

List<String> getTeamMembers(Map<String, String> teams, String team) {
  return teams.entries
      .where((e) => e.value == team)
      .map((e) => e.key)
      .toList();
}

File: server/src/engine/team-assigner.ts
import { Card, Team, Rank, Suit } from './types';

export function assignTeams(
  hands: Map<string, Card[]>
): Map<string, Team> {
  const teams = new Map<string, Team>();

  const redThreeHolders: string[] = [];
  for (const [playerId, hand] of hands) {
    const hasRedThree = hand.some(
      (c) =>
        (c.suit === Suit.HEART || c.suit === Suit.DIAMOND) &&
        c.rank === Rank.THREE
    );
    if (hasRedThree) redThreeHolders.push(playerId);
  }

  for (const [playerId] of hands) {
    teams.set(playerId, redThreeHolders.includes(playerId) ? 'red' : 'black');
  }

  return teams;
}

export function getRedThreeHolder(
  hands: Map<string, Card[]>,
  suit: Suit
): string | null {
  for (const [playerId, hand] of hands) {
    const has = hand.some(
      (c) => c.suit === suit && c.rank === Rank.THREE
    );
    if (has) return playerId;
  }
  return null;
}

export function getTeamMembers(
  teams: Map<string, Team>,
  team: Team
): string[] {
  const members: string[] = [];
  for (const [playerId, t] of teams) {
    if (t === team) members.push(playerId);
  }
  return members;
}
